File: linked-list/cpp/reverse-list-1.cpp
/* Reverse a singly linked list.
   LC 206. Reverse Linked List
   Author: Shruti
*/

//Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head || !head->next) {
            return head;
        }
        ListNode* curr = head;
        ListNode* nxt = head;
        ListNode* prev = NULL;
        while(curr) {
            nxt = curr->next;
            curr->next = prev;
            prev = curr;
            curr = nxt;
        }
        return prev;
    }
};

File: linked-list/cpp/reverse-list-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        temp_pre = None
        temp = head
        while(temp is not None):
            n = temp.next
            temp.next = temp_pre
            temp_pre = temp
            temp = n
        head = temp_pre
        return head